operations/host_spawn.go

```diff
@@ -792,3 +796,367 @@ func hostRunCommand() cli.Command {
+			makeParentDirs := c.BoolT(makeParentDirectoriesFlagName)
+			makeParentDirsOnRemote := makeParentDirs && !pull && !remoteIsLocal
+			opts := rsyncOpts{
+				local:                localPath,
+				remote:               remotePath,
+				user:                 user,
+				host:                 host,
+				makeRemoteParentDirs: makeParentDirsOnRemote,
+				excludePatterns:      c.StringSlice(excludeFlagName),
+				shouldDelete:         c.BoolT(deleteFlagName),
+				pull:                 pull,
+				dryRun:               dryRun,
+			}
+			if makeParentDirs && !makeParentDirsOnRemote {
+				if err := makeLocalParentDirs(localPath, remotePath, pull); err != nil {
+					return errors.Wrap(err, "could not create directory structure")
+				}
+			}
+			cmd, err := buildRsyncCommand(ctx, opts)
+			if err != nil {
+				return errors.Wrap(err, "could not build rsync command")
+			}
+			return cmd.Run()
+		},
+	}
+}
+
+// makeLocalParentDirs creates the parent directories for the rsync destination
+// depending on whether or not we are doing a push or pull operation.
+func makeLocalParentDirs(local, remote string, pull bool) error {
+	var dirs string
+	if pull {
+		dirs = filepath.Dir(local)
+	} else {
+		dirs = filepath.Dir(remote)
+	}
+
+	return errors.Wrapf(os.MkdirAll(dirs, 0755), "could not create local parent directories '%s'", dirs)
+}
+
+// getUserAndHostname gets the user's spawn hosts and if the hostID matches one
+// of the user's spawn hosts, it returns the username and hostname for that
+// host.
+func getUserAndHostname(ctx context.Context, hostID, confPath string) (user, hostname string, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	conf, err := NewClientSettings(confPath)
+	if err != nil {
+		return "", "", errors.Wrap(err, "problem loading configuration")
+	}
+	client := conf.setupRestCommunicator(ctx)
+	defer client.Close()
+
+	params := model.APIHostParams{
+		UserSpawned: true,
+		Mine:        true,
+	}
+	hosts, err := client.GetHosts(ctx, params)
+	if err != nil {
+		return "", "", errors.Wrap(err, "problem getting your spawn hosts")
+	}
+
+	for _, h := range hosts {
+		if model.FromStringPtr(h.Id) == hostID {
+			catcher := grip.NewBasicCatcher()
+			user = model.FromStringPtr(h.User)
+			catcher.ErrorfWhen(user == "", "could not find login user for host '%s'", hostID)
+			hostname = model.FromStringPtr(h.HostURL)
+			catcher.ErrorfWhen(hostname == "", "could not find hostname for host '%s'", hostID)
+			return user, hostname, catcher.Resolve()
+		}
+	}
+	return "", "", errors.Errorf("could not find host '%s' in user's spawn hosts", hostID)
+}
+
+// sanityCheckRsync performs some basic sanity checks for the common case in
+// which you want to mirror two directories. The trailing slash means to
+// overwrite all of the contents of the destination directory, so we check that
+// they really want to do this.
+func sanityCheckRsync(localPath, remotePath string, pull bool) (ok bool, err error) {
```

Comment:
I think that you shouldn't declare the return argument values

Reply:
I dislike the alternative, which is that you're forced to read the implementation to understand what the return values are.